crates/unladen_swallow/src/graphs/gen.rs

```diff
@@ -0,0 +1,252 @@
+use super::{load_from_py, Layout, NumpyType, SharedSlice};
+use crate::pr::PageRankResult;
+use graph::prelude::{
+    CsrLayout, DirectedDegrees, DirectedNeighbors, Graph as GraphTrait, Graph500, Idx,
+    RelabelByDegreeOp, ToUndirectedOp, UndirectedDegrees, UndirectedNeighbors,
+};
+use numpy::PyArray1;
+use pyo3::{
+    exceptions::PyValueError,
+    prelude::*,
+    types::{PyDict, PyList},
+};
+use std::{marker::PhantomData, path::PathBuf, sync::Arc, time::Duration};
+
+pub(crate) struct PyGraph<NI, G> {
+    g: Arc<G>,
+    pub(crate) load_micros: u64,
+    _ni: PhantomData<NI>,
+}
+
+impl<NI, G> PyGraph<NI, G> {
+    pub(crate) fn new(load_micros: u64, g: G) -> Self {
+        Self {
+            g: Arc::new(g),
+            load_micros,
+            _ni: PhantomData,
+        }
+    }
+
+    pub(crate) fn g(&self) -> &G {
+        &*self.g
+    }
+}
+
+/// pymethods
+impl<NI, G> PyGraph<NI, G>
+where
+    NI: Idx,
+    G: TryFrom<(Graph500<NI>, CsrLayout)> + Send,
+    graph::prelude::Error: From<G::Error>,
+{
+    /// Load a graph in the Graph500 format
+    pub(crate) fn load(py: Python<'_>, path: PathBuf, layout: Layout) -> PyResult<Self> {
+        load_from_py(py, path, layout, |g, took| Self::new(took, g))
+    }
+}
+
+/// pymethods
+impl<NI, G> PyGraph<NI, G>
+where
+    NI: Idx,
+    G: GraphTrait<NI>,
+{
+    /// Returns the number of nodes in the graph.
```

Comment:
I haven't found a good way to deduplicate the docs. We need them (or want them) on the actual implementations because that is what the untyped tooling (such as ipython and jupyter) uses and we need them on the .pyi file because that is what the typed tooling (such as IDEs) uses.

Reply:
:sad-ferris: